mgr/dashboard: Silences Page e2e Test

adk3798 opened the pull request:
Tests for correct breadcrumb on Silences pages
Tests the ability to create and expire a silence

Fixes: https://tracker.ceph.com/issues/40788
Fixes: https://tracker.ceph.com/issues/40930

Signed-off-by: Adam King <adking@redhat.com>
Signed-off-by: Rafael Quintero <rquinter@redhat.com>

tspmelo commented on src/pybind/mgr/dashboard/frontend/e2e/cluster/silences.po.ts:
```diff
@@ -0,0 +1,122 @@
+  expire(creator) {
+    this.navigateTo();
+
+    expect(this.getBreadcrumbText()).toEqual('Silences');
```
This is already tested in the breadcrumb test
```suggestion
```

tspmelo commented on src/pybind/mgr/dashboard/frontend/e2e/cluster/silences.po.ts:
```diff
@@ -0,0 +1,122 @@
+
+    $('.table-actions button.dropdown-toggle').click(); // click toggle menu
+    $('li.expire a').click(); // click expire
+    browser.wait(Helper.EC.visibilityOf($('.modal-title.float-left')), 10000).then(() => {
```
I think having just the 2nd wait is enough.
```suggestion
```

tspmelo commented on src/pybind/mgr/dashboard/frontend/e2e/cluster/silences.po.ts:
```diff
@@ -0,0 +1,122 @@
+    expect(this.getBreadcrumbText()).toEqual('Silences');
+
+    // Order silences so most recently updated one is on top
+    element(by.cssContainingText('.datatable-header-cell-label.draggable', 'Updated'))
```
As discussed before, I think we should filter the table with a specific string to be sure we are testing the correct Silence.
We can't be sure some other test won't change a Silence and that one will be the most recent.

tspmelo commented on src/pybind/mgr/dashboard/frontend/e2e/cluster/silences.po.ts:
```diff
@@ -0,0 +1,122 @@
+    $('.table-actions button.dropdown-toggle').click(); // click toggle menu
+    $('li.expire a').click(); // click expire
+    browser.wait(Helper.EC.visibilityOf($('.modal-title.float-left')), 10000).then(() => {
+      browser.wait(Helper.EC.visibilityOf($('.custom-control-label')), 5000);
```
```suggestion
      browser.wait(Helper.EC.visibilityOf($('.custom-control-label')), Helper.TIMEOUT);
```

tspmelo commented on src/pybind/mgr/dashboard/frontend/e2e/cluster/silences.po.ts:
```diff
@@ -0,0 +1,122 @@
+            Helper.EC.not(
+              Helper.EC.presenceOf(element(by.cssContainingText('.modal-dialog', 'Expire Silence')))
+            ),
+            5000
```
```suggestion
            Helper.TIMEOUT
```

tspmelo commented on src/pybind/mgr/dashboard/frontend/e2e/cluster/silences.po.ts:
```diff
@@ -0,0 +1,122 @@
+        .click()
+        .then(() => {
+          browser.wait(
+            Helper.EC.not(
```
```suggestion
```

tspmelo commented on src/pybind/mgr/dashboard/frontend/e2e/cluster/silences.po.ts:
```diff
@@ -0,0 +1,122 @@
+        .then(() => {
+          browser.wait(
+            Helper.EC.not(
+              Helper.EC.presenceOf(element(by.cssContainingText('.modal-dialog', 'Expire Silence')))
```
I think it's more clean, when possible, to use 1 EC instead of 2.
```suggestion
              Helper.EC.stalenessOf(element(by.cssContainingText('.modal-dialog', 'Expire Silence')))
```

tspmelo commented on src/pybind/mgr/dashboard/frontend/e2e/cluster/silences.po.ts:
```diff
@@ -0,0 +1,122 @@
+          Helper.EC.not(
+            Helper.EC.presenceOf(element(by.cssContainingText('.modal-dialog', 'Matcher')))
+          ),
+          5000
```
```suggestion
          Helper.TIMEOUT
```

tspmelo commented on src/pybind/mgr/dashboard/frontend/e2e/cluster/silences.po.ts:
```diff
@@ -0,0 +1,122 @@
+        // wait for modal container where you select Matcher to dissapear
+        browser.wait(
+          Helper.EC.not(
+            Helper.EC.presenceOf(element(by.cssContainingText('.modal-dialog', 'Matcher')))
```
```suggestion
            Helper.EC.stalenessOf(element(by.cssContainingText('.modal-dialog', 'Matcher')))
```

tspmelo commented on src/pybind/mgr/dashboard/frontend/e2e/cluster/silences.po.ts:
```diff
@@ -0,0 +1,122 @@
+        const createButton = element(by.cssContainingText('button', 'Create Silence'));
+        // wait for modal container where you select Matcher to dissapear
+        browser.wait(
+          Helper.EC.not(
```
```suggestion
```

tspmelo commented:
jenkins test dashboard

tspmelo commented:
jenkins test dashboard

tspmelo commented:
Not sure if prometheus is configured in jenkins... 
@callithea

callithea commented:
jenkins test dashboard

callithea commented:
> Not sure if prometheus is configured in jenkins...
> @callithea

No, I think it's not configured (at least not from our side or if prometheus is configured/enabled we would need to configure Grafana from a dashboard perspective).
Note: I just triggered the job again to test a change of the Jenkins job.

adk3798 commented:
I know there's something going on with Jenkins right now but I decided to go ahead and make some changes Tiago recommended anyway.

- Changed instances of Helper.EC.not(Helper.EC.presenceOf to Helper.EC.stalenessOf
- Changed all wait times to be Helper.TIMEOUT
- Removed breadcrumb check in expire function
- Added a few lines that put the creator into the filter box at the beginning of the expire function
- Removed wait for modal container title since waiting for the checkbox's visibility is enough

adk3798 commented:
Updated e2e to reflect changes made in #29324

callithea commented:
Hi @adk3798, I think we need to enable/configure Prometheus/Grafana within the Jenkins job in order to run this tests - I'm not sure what exactly needs to be configured (or what is already configured by default on the remote system, at least we need to make sure that Prometheus is enabled and I guess it has to be configured as data-source for Grafana) and I'm not sure if I can investigate/work on this during this week (and will be on vacation after this week). Maybe @tspmelo or @p-na can help out with that.

callithea commented:
@adk3798: Could you please apply the following lines to ``run-frontend-e2e-tests.sh`` (after line 57) within this PR so we could see if the config is sufficient to be able to run your e2e tests:

```
# Enable Prometheus module
./bin/ceph mgr module enable prometheus

# Set Alertmanager API Host
./bin/ceph dashboard set-altertmanager-api-host 'http://localhost:9093'

# Set Prometheus server
./bin/ceph dashboard set-prometheus-api-host 'http://localhost:9090'
```

alfonsomthd commented:
> # Set Alertmanager API Host
> ./bin/ceph dashboard set-altertmanager-api-host 'http://localhost:9093'
altertmanager => alertmanager

adk3798 commented:
Hello @callithea I know there are other issues with the dashboard tests right now you're addressing with #29426 but could you check that the changes to the front-end e2e script I added to this PR are what you wanted?

Thanks

callithea commented:
> > ./bin/ceph dashboard set-altertmanager-api-host 'http://localhost:9093'
> 
> altertmanager => alertmanager

Thanks @alfonsomthd, fixed the typo in my comment!

callithea commented:
> Hello @callithea I know there are other issues with the dashboard tests right now you're addressing with #29426 but could you check that the changes to the front-end e2e script I added to this PR are what you wanted?
> 
> Thanks

Hi @adk3798 looks good to me! Let's see if it works after the webdriver issue has been fixed..

adk3798 commented:
jenkins test dashboard

adk3798 commented:
jenkins test dashboard

s0nea commented:
jenkins test dashboard

adk3798 commented:
Going to remove the merge commit

Devp00l commented on src/pybind/mgr/dashboard/run-frontend-e2e-tests.sh:
```diff
@@ -55,24 +55,29 @@ if [ "$BASE_URL" == "" ]; then
+    ./bin/ceph dashboard set-prometheus-api-host 'http://localhost:9090'
 
-    BASE_URL=$(./bin/ceph mgr services | jq -r .dashboard)
+    BASE_URL=$(./bin/ceph mgr services | jq .dashboard)
```
All changes in this file beginning with line 65 seem unrelated  to the actual issue, I'm also not sure if all these changes are helpful for Jekins. If they are needed could you please put them into a different commit?

adk3798 replied:
That was actually a mistake. Only the extra lines on 58-63 lines were meant to change. Thanks for catching, I'll change it back

adk3798 commented:
Update:
 - Removed accidental changes to run-frontend-e2e-tests.sh. The only changes that should be present are the 6 extra lines from line 58-63

callithea commented:
jenkins retest this please

adk3798 commented:
> Hi @adk3798, unfortunately the dashboard tests failed with
> `E/launcher - Error: InvalidElementStateError: invalid element state: Failed to execute 'replace' on 'Location': '/#/login' is not a valid URL.`
> I can also reproduce this locally. Could you please take a look?

![Screenshot from 2019-09-04 11-17-56](https://user-images.githubusercontent.com/47704447/64268269-c3e71c00-cf05-11e9-8665-53f5d72f0cdc.png)

@callithea The most recent update seems to have gotten past that error, at least within the Jenkins test. However, the silences page tests themselves are still failing. I'm surprised that even the breadcrumb test failed as I think I remember that passing in the past. I'm going to run the dashboard tests one more time since the images page tests also failed just to make sure the results are consistent.

adk3798 commented:
jenkins test dashboard

adk3798 commented:
jenkins test dashboard

adk3798 commented:
jenkins test make check

adk3798 commented:
jenkins test dashboard

adk3798 commented:
> > Hi @adk3798, unfortunately the dashboard tests failed with
> > `E/launcher - Error: InvalidElementStateError: invalid element state: Failed to execute 'replace' on 'Location': '/#/login' is not a valid URL.`
> > I can also reproduce this locally. Could you please take a look?
> 
> ![Screenshot from 2019-09-04 11-17-56](https://user-images.githubusercontent.com/47704447/64268269-c3e71c00-cf05-11e9-8665-53f5d72f0cdc.png)
> 
> @callithea The most recent update seems to have gotten past that error, at least within the Jenkins test. However, the silences page tests themselves are still failing. I'm surprised that even the breadcrumb test failed as I think I remember that passing in the past. I'm going to run the dashboard tests one more time since the images page tests also failed just to make sure the results are consistent.

After running the dashboard tests 4 times, twice they ran normally and twice they got the 'Location': '/#/login' is not a valid URL.` error. I've been unable to recreate the error locally. Regardless, the silences page tests do not pass so I don't think the changes to the e2e script have enabled Prometheus. Do you have any ideas on what I should look at/read up on to figure out how to fix this @callithea . I'm not really sure where to start with this.

callithea commented:
> > > Hi @adk3798, unfortunately the dashboard tests failed with
> > > `E/launcher - Error: InvalidElementStateError: invalid element state: Failed to execute 'replace' on 'Location': '/#/login' is not a valid URL.`
> > > I can also reproduce this locally. Could you please take a look?
> > 
> > 
> > ![Screenshot from 2019-09-04 11-17-56](https://user-images.githubusercontent.com/47704447/64268269-c3e71c00-cf05-11e9-8665-53f5d72f0cdc.png)
> > @callithea The most recent update seems to have gotten past that error, at least within the Jenkins test. However, the silences page tests themselves are still failing. I'm surprised that even the breadcrumb test failed as I think I remember that passing in the past. I'm going to run the dashboard tests one more time since the images page tests also failed just to make sure the results are consistent.
> 
> After running the dashboard tests 4 times, twice they ran normally and twice they got the 'Location': '/#/login' is not a valid URL.` error. I've been unable to recreate the error locally. Regardless, the silences page tests do not pass so I don't think the changes to the e2e script have enabled Prometheus. Do you have any ideas on what I should look at/read up on to figure out how to fix this @callithea . I'm not really sure where to start with this.

Hi @adk3798, the added configuration for Prometheus in the e2e script should be sufficient to enable Prometheus and to make the "Silences" page visible. The best way to verify that it also works in Jenkins would be to access the system (UI) while it is running the tests, I'll try to check this.

Also I currently see a lot of e2e test failures when running the tests locally on master and in other PRs (timeouts and `No element found using locator`), so I think we currently have issues with running the e2e tests in general. :(

callithea commented:
Hi @adk3798, I'd suggest after PR#30377 has been merged and after rebasing this PR we can again take a look at the dashboard test results.

adk3798 commented:
Rebased. Sorry for the delay

callithea commented:
> Rebased. Sorry for the delay

Thanks! No worries

callithea commented:
jenkins test dashboard

callithea commented:
jenkins test dashboard

callithea commented:
jenkins test dashboard

callithea commented:
The e2e tests again fail with ``Error: InvalidElementStateError: invalid element state: Failed to execute 'replace' on 'Location': '/#/login' is not a valid URL.``
@tspmelo IIRC you also had this issue in another PR..?

tspmelo commented:
I think the problem I had before was different.
Last time I tested this PR, what was causing the problem was one of the lines in `run-frontend-e2e-tests.sh`.
If I comment it the tests would run, but the silences tests would still fail.

p-se commented on src/pybind/mgr/dashboard/frontend/e2e/cluster/silences.po.ts:
```diff
@@ -0,0 +1,100 @@
+    );
+  }
+
+  // expire is currently limited to expiring the most recently updated silence with the given creator name
```
Usually documentation of classes and methods are documented according to [JSDoc specification](https://devdocs.io/jsdoc/howto-es2015-classes):

```javascript
/**
 * Description of class
 */
class X {
  /**
   * Description of method
   */
  method() {
  }
}
```

The specification contains the recommendation to use `@param` or `@return` keywords. We discourage to use those keywords as a complete documentation can be created out of properly typed TypeScript:

```typescript
class X {
  /*
   * Description
   */
  async expire(creator: string): void {
}
```

p-se commented:
The test works fine if tested locally without running `run-frontend-e2e-tests.sh`, but the current versio of the script causes some issues.

adk3798 commented:
Would you like me to add the sleep 10 on line 61? Despite the fact that the dashboard tests will still fail in Jenkins (for now), it seems like that line will have to be added at some point anyway.

adk3798 commented:
Changed:
 - Rebased
 - Added sleep 10 after enabling prometheus module in run frontend script
 - Changed MGR=2 to MGR=1 in run frontend script

callithea commented:
> Changed:
> 
>     * Rebased
> 
>     * Added sleep 10 after enabling prometheus module in run frontend script
> 
>     * Changed MGR=2 to MGR=1 in run frontend script

JFYI: There's currently a different issue with the dashboard Jenkins job (https://tracker.ceph.com/issues/42487) I hope it can be fixed soon so we could retest your latest changes!

callithea commented:
jenkins test dashboard

callithea commented:
jenkins test dashboard

callithea commented:
@adk3798 JFYI: We're currently facing other issues with the dashboard e2e Jenkins job:
- https://tracker.ceph.com/issues/42611
- https://tracker.ceph.com/issues/42608
- https://tracker.ceph.com/issues/42610

I'll let you know as soon as we can try to test your PR again and sorry for the delay again!

callithea commented:
jenkins test dashboard

adk3798 commented:
Just rebased. Didn't notice the edit to your comment until now @callithea, my mistake.

alfonsomthd commented:
jenkins test make check

alfonsomthd commented:
jenkins test dashboard

callithea commented:
Update: We need to install some packages on the Jenkins system in order to get the tests to work.
Seems like the packages `golang-github-prometheus-alertmanager` and
`golang-github-prometheus` are not available for CentOS 7 (which is the OS that is being used for the dashboard Jenkins job).
So we need to figure out if we could set this up as a container or if we want to download the binary and install it. I need to investigate on this..

s0nea commented:
enkins test dashboard

s0nea commented:
jenkins test dashboard

callithea commented:
> Update: We need to install some packages on the Jenkins system in order to get the tests to work.
> Seems like the packages `golang-github-prometheus-alertmanager` and
> `golang-github-prometheus` are not available for CentOS 7 (which is the OS that is being used for the dashboard Jenkins job).
> So we need to figure out if we could set this up as a container or if we want to download the binary and install it. I need to investigate on this..

Hi @adk3798! I added a script ([PR is open](https://github.com/ceph/ceph-build/pull/1497)) which is supposed to take care of getting Prometheus and Alertmanager up and running on the Jenkins system, but I need some time in order to test it on Jenkins to see if some configuration for Prometheus and Alertmanager is required.

tspmelo commented:
Currently is not possible to run these tests.
After merging #34285 we can open a new PR and refactor this using fixtures.
